libsolidity/analysis/TypeChecker.cpp

```diff
@@ -1681,352 +1681,556 @@ void TypeChecker::endVisit(BinaryOperation const& _operation)
-							"Named argument \"" + *argumentNames[i] +  "\" does not match function declaration."
-						);
+					param_arg_map[i] = nullptr;
+					not_all_mapped = true;
+					m_errorReporter.typeError(
+						_functionCall.location(),
+						"Named argument \"" +
+						*argumentNames[i] +
+						"\" does not match function declaration."
+					);
 				}
+			}
+
+			if (not_all_mapped)
+				return;
+		}
+	}
+
+	// Check for compatible types between arguments and parameters
+	for (size_t i = 0; i < param_arg_map_size; ++i)
+		if (!type(*param_arg_map[i])->isImplicitlyConvertibleTo(*parameterTypes[i]))
+		{
+			solAssert(!!param_arg_map[i], "unmapped parameter");
+
+			string msg =
+				"Invalid type for argument in function call. "
+				"Invalid implicit conversion from " +
+				type(*param_arg_map[i])->toString() +
+				" to " +
+				parameterTypes[i]->toString() +
+				" requested.";
+			if (
+				_functionType->kind() == FunctionType::Kind::BareCall ||
+				_functionType->kind() == FunctionType::Kind::BareCallCode ||
+				_functionType->kind() == FunctionType::Kind::BareDelegateCall ||
+				_functionType->kind() == FunctionType::Kind::BareStaticCall
+			)
+				msg +=
+					" This function requires a single bytes argument."
+					" If all your arguments are value types, you can"
+					" use abi.encode(...) to properly generate it.";
+			else if (
+				_functionType->kind() == FunctionType::Kind::KECCAK256 ||
+				_functionType->kind() == FunctionType::Kind::SHA256 ||
+				_functionType->kind() == FunctionType::Kind::RIPEMD160
+			)
+				msg +=
+					" This function requires a single bytes argument."
+					" Use abi.encodePacked(...) to obtain the pre-0.5.0"
+					" behaviour or abi.encode(...) to use ABI encoding.";
+			m_errorReporter.typeError(param_arg_map[i]->location(), msg);
 		}
+}
+
+bool TypeChecker::visit(FunctionCall const& _functionCall)
+{
+	vector<ASTPointer<Expression const>> const& arguments = _functionCall.arguments();
+	bool argumentsArePure = true;
+
+	// We need to check arguments' type first as they will be needed for overload resolution.
+	for (ASTPointer<Expression const> const& argument: arguments)
+	{
+		argument->accept(*this);
+		if (!argument->annotation().isPure)
+			argumentsArePure = false;
+	}
+
+	// For positional calls only, store argument types
+	if (_functionCall.names().empty())
+	{
+		shared_ptr<TypePointers> argumentTypes = make_shared<TypePointers>();
+		for (ASTPointer<Expression const> const& argument: arguments)
+			argumentTypes->push_back(type(*argument));
+		_functionCall.expression().annotation().argumentTypes = move(argumentTypes);
 	}
 
-	if (returnTypes.size() == 1)
-		_functionCall.annotation().type = returnTypes.front();
+	_functionCall.expression().accept(*this);
+
+	TypePointer const expressionType = type(_functionCall.expression());
+	TypePointer actualType;
+
+	// Determine type for type conversion or struct construction expressions
+	if (expressionType->category() == Type::Category::TypeType)
```

Comment:
Checking category() first avoids the overhead of dynamic casts on all function call nodes.
I would normally use a static cast instead of dynamic here, as it appears category() == TypeType if and only if expressionType is pointing to a TypeType instance.

However as dynamic casts are used throughout TypeChecker I left it as a dynamic cast.
Happy to make it a static cast or take out the category() check - please advise.

Reply:
The choice is between dynamic cast of pointer or check plus dynamic cast of reference. Since dynamic cast of pointer is shorter, I would prefer that.

Reply:
The choices I was suggesting were either dynamic cast of pointer or check plus static cast - the second being in this case:

    if (expressionType->category() == Type::Category::TypeType)
        actualType = static_cast<TypeType const*>(expressionType.get())->actualType();

Reply:
Readability and safety go before efficiency. If you want to verify the correct relation between the runtime type and the result of `category()` then please use a condition on `category()` plus a dynamic cast of reference type (which throws if the argument's type does not match).